### DISCUSSION TITLE
Why in some cases error model is set even if request is a success?

### DISCUSSION THREAD

**Discussion by thetarby** [go-resty/resty#522] [General]
In middleware.go, the code where error is set to nil is never reached if status is 201 no content because of the if block at the beginning of the method. 
```go
func parseResponseBody(c *Client, res *Response) (err error) {
	if res.StatusCode() == http.StatusNoContent {
		return
	}
	// Handles only JSON or XML content type
	ct := firstNonEmpty(res.Request.forceContentType, res.Header().Get(hdrContentTypeKey), res.Request.fallbackContentType)
	if IsJSONType(ct) || IsXMLType(ct) {
		// HTTP status code > 199 and < 300, considered as Result
		if res.IsSuccess() {
			res.Request.Error = nil
			if res.Request.Result != nil {
				err = Unmarshalc(c, ct, res.body, res.Request.Result)
				return
			}
		}

		// HTTP status code > 399, considered as Error
		if res.IsError() {
			// global error interface
			if res.Request.Error == nil && c.Error != nil {
				res.Request.Error = reflect.New(c.Error).Interface()
			}

			if res.Request.Error != nil {
				err = Unmarshalc(c, ct, res.body, res.Request.Error)
			}
		}
	}

	return
}
```

This makes it impossible to check errors like 
```go
errM := resp.Error()
if errM == nil {
          // ...
}
```

This does not seem to be the desired behavior to me, am I missing something or is this an issue?

**Comment by kecci** [comment]
Hi @thetarby,

Before the discussion, I assume you are refer the code of middleware was get from here: https://github.com/go-resty/resty/blob/master/middleware.go#L322

As we can see, the func called `parseResponseBody` which is the something to do with `parsing response body to another thing`.

And status code `201` it's not an error, just a http status that show response the empty content.

If you want to makesure this is empty content, you can have 2 ways to validate it:

### Option 1, using `res.StatusCode() == http.StatusNoContent`:
```go
	if res.StatusCode() == http.StatusNoContent {
		return errors.New("the response is no content")
	}
```

### Option 2, using `res.Body == nil`:
```go
	if res.Body == nil {
		return errors.New("the response is no content")
	}
```

Hopefully this could be help your discussion.
